GopherCAN_buffers.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * GopherCAN_buffers.c
+ *
+ *  Created on: Jan 4, 2023
+ *      Author: jonathan
+ */
+
+#include "GopherCAN_buffers.h"
+
+static void copy_message(CAN_MSG* source, CAN_MSG* dest);
+
+// RX CAN buffer
+CAN_MSG rxbuff_mem[RX_BUFFER_SIZE];
+CAN_MSG_RING_BUFFER rxbuff = {ALL_BUSSES, NULL, rxbuff_mem, RX_BUFFER_SIZE, 0, 0, NULL};
+
+// TX buffers for each CAN bus
+CAN_MSG txbuff0_mem[TX_BUFFER_SIZE];
+CAN_MSG_RING_BUFFER txbuff0 = {GCAN0, NULL, txbuff0_mem, TX_BUFFER_SIZE, 0, 0, NULL};
+
+#if NUM_OF_BUSSES > 1
+CAN_MSG txbuff1_mem[TX_BUFFER_SIZE];
+CAN_MSG_RING_BUFFER txbuff1 = {GCAN1, NULL, txbuff1_mem, TX_BUFFER_SIZE, 0, 0, NULL};
+#endif
+
+#if NUM_OF_BUSSES > 2
+CAN_MSG txbuff2_mem[TX_BUFFER_SIZE];
+CAN_MSG_RING_BUFFER txbuff2 = {GCAN2, NULL, txbuff2_mem, TX_BUFFER_SIZE, 0, 0, NULL};
+#endif
+
+
+// get TX buffer associated with a CAN handle
+// defaults to TX buffer 0
+CAN_MSG_RING_BUFFER* choose_tx_buffer_from_hcan(CAN_HandleTypeDef* hcan) {
+#if NUM_OF_BUSSES > 2
+    if (txbuff2.hcan == hcan) return &txbuff2;
+#endif
+#if NUM_OF_BUSSES > 1
+    if (txbuff1.hcan == hcan) return &txbuff1;
+#endif
+    return &txbuff0;
+}
+
+// get CAN handle associated with buffer
+// defaults to TX buffer 0
+CAN_HandleTypeDef* choose_hcan_from_tx_buffer(CAN_MSG_RING_BUFFER* buffer) {
+#if NUM_OF_BUSSES > 2
+    if (&txbuff2 == buffer) return txbuff2.hcan;
+#endif
+#if NUM_OF_BUSSES > 1
+    if (&txbuff1 == buffer) return txbuff1.hcan;
+#endif
+    return txbuff0.hcan;
+}
+
+// get TX buffer on the CAN bus associated with a module ID
+// defaults to TX buffer 0
+CAN_MSG_RING_BUFFER* choose_tx_buffer_from_dest_module(MODULE_ID module) {
+#if NUM_OF_BUSSES > 2
+    if (txbuff2.BUS_ID == module_bus_number[module]) return &txbuff2;
+#endif
+#if NUM_OF_BUSSES > 1
+    if (txbuff1.BUS_ID == module_bus_number[module]) return &txbuff1;
+#endif
+    return &txbuff0;
+}
+
+//  will remove the first element of the ring buffer. If the buffer is empty it will do nothing
+void remove_from_front(CAN_MSG_RING_BUFFER* buffer)
+{
+    if (IS_EMPTY(buffer)) return;
+
+    // move the head to the next element
+    buffer->head = (buffer->head + 1) % buffer->size;
+
+    // decrement the fill level
+    buffer->fill_level--;
+}
+
+// add_message_by_highest_prio
+//  This function will add message to the buffer based on the ID of the message. Higher
+//  priority messages (lower ID) will be towards the front, with lower priority
+//  messages (greater ID) will be towards the back. Removing from the front will get
+//  the highest priority message.
+void add_message_by_highest_prio(CAN_MSG_RING_BUFFER* buffer, CAN_MSG* message)
+{
+    // protect buffer from RTOS thread switching
+    if (buffer->mutex != NULL) {
+        if(osMutexAcquire(buffer->mutex, MUTEX_TIMEOUT)) return;
+    }
+#if TARGET == F7XX || TARGET == F4XX
+    // protect buffer from interrupts
+    HAL_CAN_DeactivateNotification(choose_hcan_from_tx_buffer(buffer), CAN_IT_TX_MAILBOX_EMPTY);
```

Comment:
Tbh I wasn't sure how interrupts interacted with threads. osMutexAcquire claims the mutex for whatever task is active. But in the interrupt, which task would that be?

Reply:
The GCAN hardware handling task calls this function, and also all of the TX interrupts. This is to maximize the amount of messages that are sent in a short amount of time. Talk to sebastian about the best way to do this, he knows it better than me

Reply:
Looked into it some more, I think it might be impossible to prevent the ISR from preempting the Task without disabling the interrupt. The mutex might be worthless then, I am going to think on it some more

Reply:
My thought process was that the mutex would prevent collisions between a task calling send_parameter/send_command/etc and a different task calling service_tx.

Disabling the interrupt while writing to the buffer prevents the interrupts from attempting a second write in the middle of the first one.

Reply:
Ya thats what its for, I am just saying the interrupt disabling should be enough and the mutex doesnt do anything without stopping the interrupt

Reply:
On the interrupt side, I think temporarily disabling is our only option.

For thread safety, the mutex needs to move. osMutexAcquire returns an error if it's called within an interrupt. Maybe a wrapper function around service_can_tx_hardware would work? RTOS projects can call the wrapper which acquires the mutex, but interrupts just call service_can_tx_hardware directly.

Reply:
Thats a good idea. Then the wrapper function should disable the interrupt as well as we dont want to disable the interrupt in the interrupt